app/components/graph-editor/drawable-interfaces.ts

```diff
@@ -0,0 +1,398 @@
+// File: drawable-interfaces.ts
+// Created by: CJ Dimaano
+// Date created: January 9, 2016
+//
+// THIS FILE IS INTENDED TO BE IMPORTED ONLY INTO graph-editor.component.ts
+//
+
+
+import { EDGE_DEFAULTS, NODE_DEFAULTS } from "./constants";
+import * as MathEx from "./math";
+
+
+// Functions ///////////////////////////////////////////////////////////////////
+
+
+/**
+ * isDrawableEdge  
+ *   Typeguard for drawable edges.
+ */
+export function isDrawableEdge(obj: any): obj is DrawableEdge {
+    return isItThat(obj, EDGE_DEFAULTS);
+}
+
+/**
+ * isDrawableNode  
+ *   Typeguard for drawable nodes.
+ */
+export function isDrawableNode(obj: any): obj is DrawableNode {
+    return isItThat(obj, NODE_DEFAULTS);
+}
+
+/**
+ * isItThat  
+ *   Checks if `it` has the same members as `that`.
+ */
+function isItThat(it: any, that: any): boolean {
+    if (!it)
+        return false;
+    let result = true;
+    for (let p in that) {
+        result =
+            result &&
+            (p in it) &&
+            (typeof that[p] === typeof it[p]);
+        if (!result)
+            return false;
+    }
+    return result;
+}
+
+/**
+ * isEdgeOverlapped  
+ *   Checks if two edges are overlapping.  
+ *   It is assumed that the edge being checked does not have the same source
+ *   and destination nodes.
+ */
+export function isEdgeOverlapped(e: DrawableEdge, nodeEdges: Map<DrawableNode, Set<DrawableEdge>>): boolean {
```

Comment:
We should probably not allow users to try and connect two nodes with the same source and destination either by error message or by over writing the old edge. Ordinarily, I would say we should allow them to make such a thing, but this would violate the mathematical properties of graphs and does not seem to be a common use case.